networkx/lazy_imports.py

```diff
@@ -0,0 +1,179 @@
+import importlib.util
+import types
+import os
+import sys
+import pytest
+
+
+__all__ = ["lazy_import", "lazy_package_import", "lazy_importorskip"]
+
+
+def lazy_package_import(module_name, submodules=None, submod_attrs=None):
+    """Install lazily loaded submodules, and functions or other attributes.
+
+    Typically, modules import submodules and attributes as follows::
+
+      import mysubmodule
+      import anothersubmodule
+      from .foo import someattr
+
+    The idea of the lazy package import is to replace the `__init__.py`
+    module's `__getattr__`, `__dir__`, and `__all__` attributes such that
+    all imports work exactly the way they normally would, except that the
+    actual import is delayed until the resulting module object is first used.
+
+    The typical way to call this function, replacing the above imports, is::
+
+      __getattr__, __lazy_dir__, __all__ = install_lazy(
+          __name__,
+          ['mysubmodule', 'anothersubmodule'],
+          {'foo': 'someattr'}
+      )
+
+    This functionality requires Python 3.7 or higher.
+
+    Parameters
+    ----------
+
+    module_name : str
+        Typically use __name__.
+    submodules : set
+        List of submodules to lazily import.
+    submod_attrs : dict
+        Dictionary of submodule -> list of attributes / functions.
+        These attributes are created the first time the submodule
+        is used.
+
+    Returns
+    -------
+
+    __getattr__, __dir__, __all__
+
+    """
+    if submod_attrs is None:
+        submod_attrs = {}
+
+    if submodules is None:
+        submodules = set()
+    else:
+        submodules = set(submodules)
+
+    attr_to_modules = {
+        attr: mod for mod, attrs in submod_attrs.items() for attr in attrs
+    }
+
+    __all__ = list(submodules | attr_to_modules.keys())
+
+    def __getattr__(name):
+        if name in submodules:
+            return importlib.import_module(f"{module_name}.{name}")
+        elif name in attr_to_modules:
+            submod = importlib.import_module(f"{module_name}.{attr_to_modules[name]}")
+            return getattr(submod, name)
+        else:
+            raise AttributeError(f"No {module_name} attribute {name}")
+
+    def __dir__():
+        return __all__
+
+    if os.environ.get("EAGER_IMPORT", ""):
+        for attr in set(attr_to_modules.keys()) | submodules:
+            __getattr__(attr)
+
+    return __getattr__, __dir__, list(__all__)
+
+
+class DelayedImportErrorLoader(importlib.util.LazyLoader):
+    def exec_module(self, module):
+        """Make the module load lazily."""
+        module.__spec__.loader = self.loader
```

Comment:
Curious where this came from; is this the part you said you took from LazyLoader?

Reply:
I guess what I'm curious about is whether you can't:

```python
super().exec_module(module)
module.__class__ = DelayedImportErrorModule
```

Reply:
Yes!  I should use `super` and then fix the class.  Nice!